dex/elrond_dex_farm/src/lib.rs

```diff
@@ -343,29 +365,130 @@ pub trait Farm:
             .into())
     }
 
+    #[payable("*")]
+    #[endpoint(compoundRewards)]
+    fn compound_rewards(
+        &self,
+        #[payment_token] payment_token_id: TokenIdentifier,
+        #[payment_amount] payment_amount: Self::BigUint,
+        #[payment_nonce] payment_token_nonce: Nonce,
+        #[var_args] opt_accept_funds_func: OptionalArg<BoxedBytes>,
+    ) -> SCResult<CompoundRewardsResultType<Self::BigUint>> {
+        require!(self.is_active(), "Not active");
+        require!(payment_amount > 0, "Zero amount");
+
+        require!(!self.farm_token_id().is_empty(), "No issued farm token");
+        let farm_token_id = self.farm_token_id().get();
+        require!(payment_token_id == farm_token_id, "Unknown farm token");
+
+        let farming_token = self.farming_token_id().get();
+        let reward_token = self.reward_token_id().get();
+        require!(
+            farming_token == reward_token,
+            "Farming token differ from reward token"
+        );
+        self.generate_aggregated_rewards(&reward_token);
+
+        let current_rps = self.reward_per_share().get();
+        let farm_attributes = self.get_farm_attributes(&payment_token_id, payment_token_nonce)?;
+        let reward = self.calculate_reward(
+            &payment_amount,
+            &current_rps,
+            &farm_attributes.reward_per_share,
+        );
+
+        if reward > 0 {
+            self.decrease_reward_reserve(&reward)?;
+        }
+
+        let farm_token_id = self.farm_token_id().get();
+        let mut new_farm_contribution =
+            &payment_amount + &(&reward * &(farm_attributes.apr_multiplier as u64).into());
+
+        let new_initial_farming_amount = self.rule_of_three_non_zero_result(
+            &payment_amount,
+            &farm_attributes.current_farm_amount,
+            &farm_attributes.initial_farming_amount,
+        )?;
+        let new_compound_reward_amount = self.rule_of_three(
+            &payment_amount,
+            &farm_attributes.current_farm_amount,
+            &farm_attributes.compounded_reward,
+        ) + reward;
+
+        let mut new_attributes = FarmTokenAttributes {
+            reward_per_share: current_rps,
+            entering_epoch: self.blockchain().get_block_epoch(),
```

Comment:
I noticed that this field is `farm_attributes.entering_epoch` in the claimRewards. How about here?

Reply:
Here it should be like this. I'm not a fan of resetting in some instances and not resetting in others

Reply:
So if a user compounds the rewards after a year and exits the farm. Then the all the rewards will be locked after a year?

Reply:
Pretty much yes

Reply:
Humm. I wonder whether users will like this compound feature or not. Bascially, each time users compound their reward, they are delaying the time to unlock the compound rewards.

Reply:
FWIW, to my untrained eye it seems that it could have been a merge between `(payment_amount, farm_attributes.entering_epoch)` and `(reward*multiplier, get_block_epoch)`, similar to what would have happened if you would generate a new farm token for `reward*multiplier`, then you merged that with the existing entry.

Reply:
Merging with the existing entry changed similary